text/0000-forbidden-value-trait.md

```diff
@@ -0,0 +1,70 @@
+- Start Date: 2020-03-24
+- RFC PR #:
+- Rust Issue #:
+
+# Summary
+
+Some types have one or more forbidden values. Such values can be sometimes used to perform [optimizations](https://github.com/rust-lang/rust/pull/45225). For example, rustc currently uses a known forbidden value of a certain type `A` (like `&T`, `Box<T>`, `NonNull<T>`, `NonZeroI8` and some enum types) to represent the `None` enum variant of `Option<A>`. I propose to give users the ability to enable the same optimizations for types rustc currently doesn't know a forbidden value exists for.
+
+# Motivation
+
+This would increase the number of situations where the compiler can do the aforementioned, genuinely useful optimizations.
+
+# Detailed design
+
+Add the following trait somewhere in the standard library:
+```rust
+unsafe trait ForbiddenValues<A, B>
+where
+    A: FixedSizeArray<B>,
+    B: FixedSizeArray<u8>,
+{
+    const FORBIDDEN_VALUES: A;
+}
+```
+
+To implement `ForbiddenValue<A, B>` for type `T`, the following conditions must be met:
+1) Type `T` has a stable layout and representation
+2) Type `B` is a fixed-size array of `u8`
+3) Type `A` is a fixed-size array of `B`
+4) `std::mem::size_of::<T>() == std::mem::size_of::<B>()`
+5) For each value `v` of type `T`, and for each item `f` in `T::FORBIDDEN_VALUES`, `*(&v as *const T as *const B) != f`
+
+Then compilers would be allowed to use `T::FORBIDDEN_VALUES` to represent forbidden values of type `T` in whatever optimizations they decide to perform.
+
+Users could then, for example, implement a wrapper for floating point values that are always finite like this:
+```rust
+#![feature(const_transmute)]
+
+#[repr(transparent)]
+struct FastFloat(f32);
+
+unsafe impl ForbiddenValues<[[u8; 4]; 3], [u8; 4]> for FastFloat {
+    const FORBIDDEN_VALUES: [[u8; 4]; 3] = unsafe {
+        [
+            mem::transmute(f32::NAN),
```

Comment:
Note that `f32` has about 16.7 million different `NaN` values.

Reply:
No problem. You just write a procedural macro to produce those 16.8 million lines of code and buy a supercomputer to compile it 😜

I concede that your suggestion for a trait with a `const fn is_forbidden(&self) -> bool` method would be strictly better than this RFC - it would make it (relatively) easy to specify large ranges of forbidden values. But I do wonder about its compiler implementation - without knowing much at all, I would imagine that the compiler would have to call `value.is_forbidden()` for each possible bit pattern of `value`.

Reply:
Why would the compiler have to call `value.is_forbidden()` for each possible bit pattern?

If the compiler "needs" 5 values for a variety of optimizations, it would just have to find 5 forbidden values and use those.

Reply:
That would always have a worst case of going through every value, though. Which isn't scalable past a few bits.

Reply:
@clarfon Do you mean ranges as in something like the following?
```rust
unsafe trait ForbiddenValues<A, B>
where
    A: FixedSizeArray<(B, B)>,
    B: FixedSizeArray<u8>,
{
    const FORBIDDEN_VALUES: A;
}
```
Where `B` is now interpreted as a fixed-size unsigned integer type with an arbitrary width, and the tuple `(B, B)` represents the inclusive start and end of the range of forbidden values.

Reply:
To eliminate dead code, all `rustc` needs to do is insert and constant-fold the mentioned `unreachable_unchecked` snippet. This is something it can do today.

Finding holes for compactly encoding `enum` variants, however, is more tricky.